### PR TITLE
update logic to inject FastForwardSampler / CaptureIterableDataset 2/n

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

Should be merged after https://github.com/PyTorchLightning/pytorch-lightning/pull/8364

### Does your PR introduce any breaking changes ? If yes, please list them.
* Could cause different behaviour depending on the seed of samplers within the iterable dataset.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [x] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/trainer/data_loading.py [hunk 1]
```diff
@@ -159,16 +172,28 @@ def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningS
...
+                fast_forward_sampler = batch_sampler = FastForwardSampler(batch_sampler)
+
             dl_args['batch_sampler'] = batch_sampler
             dl_args['batch_size'] = 1
             dl_args['shuffle'] = False
             dl_args['sampler'] = None
             dl_args['drop_last'] = False
         else:
+            if should_use_forward_sampler:
+                fast_forward_sampler = sampler = FastForwardSampler(sampler)
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:186]:
```suggestion
                sampler = FastForwardSampler(sampler)
```

File: pytorch_lightning/trainer/data_loading.py [hunk 2]
```diff
@@ -159,16 +172,28 @@ def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningS
...
+
             dl_args['sampler'] = sampler
             dl_args['shuffle'] = False
             dl_args['batch_sampler'] = None
 
+        if should_use_forward_sampler:
+            # the forward sampler need to be informed about
+            batch_size = dl_args["batch_size"]
+            fast_forward_sampler.setup(batch_size)
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:195]:
```suggestion
            sampler.setup(batch_size)
```

File: tests/utilities/test_auto_restart.py [hunk 3]
```diff
@@ -893,3 +903,59 @@ def create_dataloader():
 
     state_dict = dataloader_to_state_dict(dataloader, iter_dataloader)
     assert state_dict == {'num_workers': 0, 'previous_worker': None, 0: {'current_iteration': 24}}
+
+
+@pytest.mark.parametrize("use_fault_tolerant", ['0', '1'])
+@RunIf(min_torch="1.6.0")
```
Comment (justusschock) [inline tests/utilities/test_auto_restart.py:909]:
```suggestion
@RunIf(min_torch="1.7.0")
```

File: pytorch_lightning/trainer/data_loading.py [hunk 4]
```diff
@@ -212,6 +241,19 @@ def __init__(self, num_features, dataset, *args, **kwargs):
 
         if not contains_dataset:
             dl_args.pop('dataset')
+        else:
+            # wrap the ``IterableDataset`` into a ``CaptureIterableDataset`` to record sampler states.
+            if fault_tolerant_enabled() and isinstance(dl_args["dataset"], IterableDataset):
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:246]:
```suggestion
            if _fault_tolerant_enabled() and isinstance(dl_args["dataset"], IterableDataset):
```

File: pytorch_lightning/trainer/data_loading.py [hunk 5]
```diff
@@ -212,6 +241,19 @@ def __init__(self, num_features, dataset, *args, **kwargs):
...
+                # force the seed in the ``Dataset``
+                seed = int(os.getenv("PL_GLOBAL_SEED", 0)) + self.current_epoch + self.global_rank
+
+                if dl_args.get("generator") is None:
+                    dl_args["generator"] = torch.Generator().manual_seed(seed)
+
+                dl_args["dataset"] = CaptureIterableDataset(
+                    dataset=dl_args["dataset"],
+                    initial_seed=dl_args["generator"].initial_seed(),
+                )
```
Comment (awaelchli) [inline pytorch_lightning/trainer/data_loading.py]: need to be careful here
loading of seed state is not yet implemented here

Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py]: Yes, definitely a weak spot.

File: pytorch_lightning/trainer/data_loading.py [hunk 6]
```diff
@@ -212,6 +241,19 @@ def __init__(self, num_features, dataset, *args, **kwargs):
 
         if not contains_dataset:
             dl_args.pop('dataset')
+        else:
+            # wrap the ``IterableDataset`` into a ``CaptureIterableDataset`` to record sampler states.
+            if _fault_tolerant_enabled() and isinstance(dl_args["dataset"], IterableDataset):
+                # force the seed in the ``Dataset``
+                seed = int(os.getenv("PL_GLOBAL_SEED", 0)) + self.current_epoch + self.global_rank
```
Comment (justusschock) [inline pytorch_lightning/trainer/data_loading.py:248]:
```suggestion
                seed = int(os.getenv("PL_GLOBAL_SEED", 0)) + self.global_rank
```

With current_epoch this might be not reproducible

File: pytorch_lightning/trainer/data_loading.py [hunk 7]
```diff
@@ -139,17 +143,26 @@ def auto_add_sampler(
...
         return dataloader
 
     @staticmethod
-    def _resolve_batch_sampler(dl_args, dataloader, sampler, mode: Optional[RunningStage] = None) -> Dict[str, Any]:
+    def _resolve_batch_sampler(
+        dl_args,
+        dataloader,
+        sampler,
+        mode: Optional[RunningStage] = None,
+        should_use_forward_sampler: Optional[bool] = False
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:161]: Is this an parameter we will want when fault tolerance is enabled by default? If not, it should be removed

File: pytorch_lightning/trainer/data_loading.py [hunk 8]
```diff
@@ -212,6 +241,19 @@ def __init__(self, num_features, dataset, *args, **kwargs):
 
         if not contains_dataset:
             dl_args.pop('dataset')
+        else:
+            # wrap the ``IterableDataset`` into a ``CaptureIterableDataset`` to record sampler states.
+            if _fault_tolerant_enabled() and isinstance(dl_args["dataset"], IterableDataset):
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:246]: We know that `has_iterable_dataset==True`

```suggestion
            if _fault_tolerant_enabled():
```

File: pytorch_lightning/trainer/data_loading.py [hunk 9]
```diff
@@ -212,6 +241,19 @@ def __init__(self, num_features, dataset, *args, **kwargs):
 
         if not contains_dataset:
             dl_args.pop('dataset')
+        else:
```
Comment (carmocca) [inline pytorch_lightning/trainer/data_loading.py:244]: Should this be `if contains_iterable_dataset`? And remove the modification to the `if` above

File: pytorch_lightning/utilities/auto_restart.py [hunk 10]
```diff
@@ -124,6 +125,15 @@ def load_state_dict(self, state_dict: Dict[int, Any], workers_initialized: bool
         self._current_iteration = state_dict[self.worker_id]["current_iteration"]
         self.restarting = True
 
+    def _get_rng_states(self):
+
+        def _collect(gen: torch.Generator):
+            return gen.get_state()
+
+        states = apply_to_collection(self.__dict__, torch.Generator, _collect)
+        states["__global"] = torch.get_rng_state()
```
Comment (carmocca) [inline pytorch_lightning/utilities/auto_restart.py:134]: why "global"?

File: tests/utilities/test_auto_restart.py [hunk 11]
```diff
@@ -893,3 +903,59 @@ def create_dataloader():
...
+            if use_fault_tolerant == '1':
+                assert isinstance(loaders["a"][0].loader.dataset, CaptureIterableDataset)
+                assert isinstance(loaders["a"][1].loader.sampler, FastForwardSampler)
+                assert isinstance(loaders["a"][2].loader.batch_sampler, FastForwardSampler)
+                assert isinstance(loaders["b"].loader.dataset, CaptureIterableDataset)
+            else:
+                assert isinstance(loaders["a"][0].loader.dataset, RangeIterableDataset)
+                assert isinstance(loaders["a"][1].loader.sampler, SequentialSampler)
+                assert isinstance(loaders["a"][2].loader.batch_sampler, CustomBatchSampler)
+                assert isinstance(loaders["b"].loader.dataset, RangeIterableDataset)
```
Comment (carmocca) [inline tests/utilities/test_auto_restart.py:955]: Not sure if these are worth adding considering they will be removed soon

File: pytorch_lightning/trainer/data_loading.py (hunk 6 above)
Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:248]: Without the `current_epoch`, the sampling would be the same if `reload_every_epochs` is True. Also, with progress tracking, we would have access to `current_epch`, so I believe it could be reproducible.

File: tests/utilities/test_auto_restart.py (hunk 11 above)
Comment (tchaton) [reply tests/utilities/test_auto_restart.py:955]: Why would they be removed ?

File: pytorch_lightning/utilities/auto_restart.py (hunk 10 above)
Comment (awaelchli) [reply pytorch_lightning/utilities/auto_restart.py:134]: the global rng state
this code is unused, we are trying to implement capturing the rng state. 
this is wip

Comment (carmocca) [reply pytorch_lightning/utilities/auto_restart.py:134]: kk, meant to ask "why global and not rng_state"

File: tests/utilities/test_auto_restart.py (hunk 11 above)
Comment (carmocca) [reply tests/utilities/test_auto_restart.py:955]: I mean when the fault tolerance experimental flag is removed :)

File: pytorch_lightning/trainer/data_loading.py (hunk 6 above)
Comment (carmocca) [reply pytorch_lightning/trainer/data_loading.py:248]: The implementation we have for `pl_worker_init_function` function does not take the `current_epoch` too so I guess suffers from the same problem. Maybe this piece could reuse that function if we change it to return the seed where `worker_id` is not used if `None`

Even if we always have the current_epoch set, the user would need to modify it before running this code to mimic an older seed.

Not sure what would be the best solution here.

File: pytorch_lightning/utilities/auto_restart.py (hunk 10 above)
Comment (justusschock) [reply pytorch_lightning/utilities/auto_restart.py:134]: all the entries in this dict will be rng states from different generators, so for my intuition it does not make sense to include `rng_state` to everyone :) Same goes for the global state :)

File: pytorch_lightning/utilities/auto_restart.py [hunk 12]
```diff
@@ -60,35 +66,37 @@ def worker_id(self) -> int:
...
-                    self.load_state_dict(self._cached_state_dict, workers_initialized=True)
-                    self._cached_state_dict = None
-
-                # when the current index matching the current_iteration, we have "fast forwarded" the sampler.
-                if self._current_iteration <= i:
-                    if self._current_iteration == i:
-                        self.restarting = False
-                    self._current_iteration += 1
-                    yield batch
+        # the `state dict` was cached as workers were available before.
```
Comment (tchaton) [inline pytorch_lightning/utilities/auto_restart.py:69]:
```suggestion
        # the `state dict` was cached as workers were unavailable before.
```

**Comment by awaelchli** [comment]
@tchaton @justusschock I removed the WIP random state restart work that we had here. I will push it into a separate PR and base it onto this one.

File: pytorch_lightning/utilities/imports.py [hunk 13]
```diff
@@ -107,4 +107,4 @@ def _fault_tolerant_enabled() -> bool:
     EXPERIMENTAL
     the `reset` function from `_MultiProcessingDataLoaderIter` was introduced in PyTorch 1.7 but we need to mock it.
     """
-    return _TORCH_GREATER_EQUAL_1_7 and bool(os.getenv("PL_FAULT_TOLERANT_TRAINING", False))
+    return _TORCH_GREATER_EQUAL_1_7 and int(os.getenv("PL_FAULT_TOLERANT_TRAINING", 0))
```
Comment (awaelchli) [inline pytorch_lightning/utilities/imports.py:110]: took me a while to figure this one out. 
beware: `bool("0") is True`

Comment (awaelchli) [reply pytorch_lightning/utilities/imports.py:110]: I added a test for this function, see below in `test_auto_restart.py`

Comment (carmocca) [reply pytorch_lightning/utilities/imports.py:110]: My bad! I broke it in a previous PR then

File: pytorch_lightning/trainer/data_loading.py [hunk 14]
```diff
@@ -223,6 +246,23 @@ def replace_sampler(self, dataloader: DataLoader, sampler, mode: Optional[Runnin
                     f"`{dataloader_cls_name}(dataset, sampler=DistributedSampler(dataset))`."
                 )
 
+        # wrap the ``IterableDataset`` into a ``CaptureIterableDataset`` to record sampler states.
+        if _fault_tolerant_enabled() and isinstance(dl_kwargs["dataset"], IterableDataset):
+            # force the seed in the ``Dataset``
+            seed = int(os.getenv("PL_GLOBAL_SEED", 0)) + self.current_epoch + self.global_rank
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:252]: Do we want to keep this ?

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:252]: potentially saver to remove until we have a good POC with random state management for restarting.

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py:252]: from an earlier discussion: https://github.com/PyTorchLightning/pytorch-lightning/pull/8366#discussion_r668056865
thoughts @justusschock @carmocca ?

**Review by awaelchli — APPROVED** [review]

File: pytorch_lightning/trainer/data_loading.py [hunk 15]
```diff
@@ -223,6 +239,16 @@ def replace_sampler(self, dataloader: DataLoader, sampler, mode: Optional[Runnin
                     f"`{dataloader_cls_name}(dataset, sampler=DistributedSampler(dataset))`."
                 )
 
+        # wrap the `IterableDataset` into a `CaptureIterableDataset` to record sampler states.
+        if _fault_tolerant_enabled() and isinstance(dl_kwargs["dataset"], IterableDataset):
+            initial_seed = dl_kwargs["generator"].initial_seed() if dl_kwargs.get("generator") is not None else None
```
Comment (tchaton) [inline pytorch_lightning/trainer/data_loading.py:244]: The initial seed doens't seem to be used anymore

Comment (awaelchli) [reply pytorch_lightning/trainer/data_loading.py]: removed

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/utilities/auto_restart.py [hunk 16]
```diff
@@ -60,35 +61,45 @@ def worker_id(self) -> int:
...
+        self._load_cached_state()
+
+        i = 0
+        sampler_iter = iter(self._sampler)
+        while i < self._current_iteration:
+            next(sampler_iter)
+            i += 1
+
+        # here: i == self._current_iteration
+        # recreate iterator to be sure loading is reflected there as well
```
Comment (justusschock) [inline pytorch_lightning/utilities/auto_restart.py:75]:
```suggestion
```
outdated; not done anymore

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/data_loading.py [hunk 17]
```diff
@@ -114,33 +119,34 @@ def auto_add_worker_init_fn(self, dataloader: DataLoader) -> None:
...
                 )
-
-            # replace with distributed sampler
             sampler = self._get_distributed_sampler(dataloader, shuffle, mode=mode)
-            dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
+        else:
+            # use current sampler
+            sampler = dataloader.sampler
+
+        dataloader = self.replace_sampler(dataloader, sampler, mode=mode)
```
Comment (yifuwang) [inline pytorch_lightning/trainer/data_loading.py:149]: Hi @tchaton, should we have replaced the sampler even when `self.accelerator_connector.replace_sampler_ddp` is `False`? I'm not familiar with the logic here, but a few of our tests failed because `replace_sampler` is now invoked after this change. Specifically, the test uses a custom sampler, but `_resolve_batch_sampler` expects to be able to re-construct the sampler following way:
```python
if (batch_sampler is not None and type(batch_sampler) is not BatchSampler) or is_predicting:
    batch_sampler = type(batch_sampler)(
        sampler,
        batch_size=batch_sampler.batch_size,
        drop_last=(False if is_predicting else batch_sampler.drop_last),
    )
```
However, the custom sampler in question doesn't have the expected fields.

Comment (tchaton) [reply pytorch_lightning/trainer/data_loading.py:149]: It won't actually replace it if `replace_sampler_ddp`. But the code should be refactored as it does more than just replace the sampler.
